README.md
[![CircleCI](https://circleci.com/gh/mozilla-services/iprepd-nginx?style=svg)](https://circleci.com/gh/mozilla-services/iprepd-nginx)

# iprepd-nginx module

`iprepd-nginx` is an openresty module for integrating with [iprepd](https://github.com/mozilla-services/iprepd).

You can use the example configuration in this repo for a standalone proxy or install using [opm](https://github.com/openresty/opm)
and integrate it yourself.

*Note:* If nginx is behind a load balancer, make sure to use something like
[ngx_http_realip_module](https://nginx.org/en/docs/http/ngx_http_realip_module.html).

## What exactly does iprepd-nginx do?

By using the `iprepd` client in `iprepd-nginx`, you can configure nginx to check the reputation of an incoming client IP within `iprepd`. With
this reputation, `iprepd-nginx` will attach up to three HTTP headers on the request that is then forwarded to your application and can reject
requests that are below the configured threshold.

These three headers are:

| Header | Values | Description |
|---|---|---|
| X-Foxsec-IP-Reputation | int (0-100) | Reputation score returned by iprepd |
| X-Foxsec-IP-Reputation-Below-Threshold | boolean ('true'/'false') | Whether the reputation is below the configured threshold |
| X-Foxsec-Block | boolean ('true'/'false') | High-level whether the request should be blocked (subject to change on what this means) |

As well, `iprepd-nginx` is designed to fail open and prefer performance to accuracy. The preference of performance to accuracy can be changed a bit as an
operator, but only to a certain extent (discussed further below).

## (Mozilla-specific) Architecture Diagram

![Architecture Diagram](docs/moz-architecture-diagram.png)

## Installation

Install using [opm](https://github.com/openresty/opm)

```
opm get mozilla-services/iprepd-nginx
```

## Operators Guide

### Prerequisites

* [iprepd](https://github.com/mozilla-services/iprepd), preferably near your `iprepd-nginx` servers (e.g. within the same region in AWS or GCP)
* A mechanism for updating iprepd. At Mozilla, this is done by feeding logs from your load balancer, application server, and potentially other locations into our [fraud detection pipeline](https://github.com/mozilla-services/foxsec-pipeline).
* (optional) A mechanism for collecting statsd metrics.

### Note on Performance

A core requirement for iprepd-nginx is that it will add no more than 10ms of latency to requests. Of the mechanisms in place to accomplish this, as an operator there are a few you should be aware of:

#### Heavy use of caching of responses from iprepd
By default, iprepd-nginx will cache all non-error responses from iprepd for 30 seconds. It is a good idea to cache errors in production, which is done by enabling `cache_errors` (discussed further below). As well, you may want to lengthen the cache ttl.

#### Strict timeouts to iprepd
By default, iprepd-nginx’s request to iprepd will timeout after 10ms. This should not be increased in production, and may be worth reducing if the network design can support it.

### Configuration of the client

#### `threshold` parameter

The `threshold` value in the client is the numerical value inbetween 0 and 100 where clients will be blocked if their
IP's reputation in iprepd is below this value.

What you will want this value to be set to will be highly contextual to your application and environment, with considerations
of what kind of violations exist, how likely a client is to activate these violations, how often a client will retry, etc.

A decent value to start at is `50`, but you will want to make sure this is tested along side the implemented iprepd
violations for your environment.

#### Audit parameters

The `audit` values in the client allow configuration of recording blocked requests to the log for security auditing purposes. The audit message includes the request body if small enough to be buffered in memory or the name of the body file if it had to be written to disk. This is dependent on the nginx setting `client_body_buffer_size`.

`audit_blocked_requests` enables auditing, by default it is disabled.

`audit_include_headers` includes the request headers in the audit message. By default, it is disabled. When enabled, Authorization and Proxy-Authorization header values are removed.

`audit_uri_list` is a table of uris that should be audited. This is mandatory when auditing is enabled. Uri's can be either a simple string or a Lua pattern.
For example,  `audit_uri_list = {"test", "/test/%d/somethingelse"}` would allow cause requests to `/test/1/somethingelse` as well as `/test` to be recorded.

#### Example

```lua
-- Parameters within options:
--  Required parameters:
--    api_key - An active API key for authenticating to iprepd
--    threshold - The reputation threshold, where IP's with a reputation below this number will
--                be blocked. There is no default for this, as it will be application specific,
--                but as described above 50 is a good starting place.
--
--  Optional parameters:
--    url - The base URL to iprepd (defaults to "http://localhost:8080/")
--    timeout - The timeout for making requests to iprepd in milliseconds (defaults to 10)
--    cache_ttl - The iprepd response cache ttl in seconds (defaults to 60)
--    cache_buffer_count - Max number of entries allowed in the cache. (defaults to 5000)
--    cache_errors - Enables (1) or disables (0) caching errors. Caching errors is a good
--                   idea in production, as it can reduce the average additional latency
--                   caused by this module if anything goes wrong with the underlying
--                   infrastructure. (defaults to disabled)
--    cache_errors_ttl - The iprepd response cache ttl for error responses (not 200 or 404) in seconds (defaults to 10)
--    statsd_host - Host of statsd collector. Setting this will enable statsd metrics collection
--    statsd_port - Port of statsd collector. (defaults to 8125)
--    statsd_max_buffer_count - Max number of metrics in buffer before metrics should be submitted
--                              to statsd (defaults to 100)
--    statsd_flush_timer - Interval for attempting to flush the stats in seconds. (defaults to 5)
--    blocking_mode - Enables (1) or disables (0) blocking within nginx by returning a
--                    403. (defaults to disabled)
--    verbose - Enables (1) or disables (0) verbose logging. Messages are logged with a
--              severity of "ERROR" so that nginx log levels do not need to be changed. (defaults
--              to disabled)
--    whitelist - List of whitelisted IP's and IP CIDR's. (defaults to empty)
-- audit_blocked_requests - records the body and optionally headers of requests that are being blocked within nginx (defaults disabled, if enabled the uris to audit must be given)
-- audit_include_headers - if audit_blocked_requests is enabled, also record the headers (defaults disabled)
-- audit_uri_list - a list of endpoints that will be audited if audit_blocked_requests is enabled (defaults to empty)
--
client = require("resty.iprepd").new({
  api_key = os.getenv("IPREPD_API_KEY"),
  threshold = 50,
  url = "http://127.0.0.1:8080",
  timeout = 10,
  cache_ttl = 30,
  cache_buffer_count = 1000,
  cache_errors = 1,
  cache_errors_ttl = 10,
  statsd_host = "127.0.0.1",
  statsd_port = 8125,
  statsd_max_buffer_count = 100,
  statsd_flush_timer = 10,
  blocking_mode = 0,
  verbose = 0,
  whitelist = "127.0.0.1,10.10.10.0/24,192.168.0.16/16",
  audit_blocked_requests = 0,
  audit_include_headers = 0,
  audit_uri_list = "/uri1,/uri2",
})
```


### Metrics (statsd)

#### Metrics that are collected

| name | type | description |
|---|---|---|
| iprepd.status.below_threshold | count | The reputation for the client ip is below the configured threshold. |
| iprepd.status.rejected | count | The request was blocked (won’t be sent if `blocking_mode` is disabled). |
| iprepd.status.accepted | count | The request was accepted. The reputation can still be below the threshold if `blocking_mode` is disabled.
| iprepd.get_reputation | count | Request to iprepd |
| iprepd.cache_hit | count | Got reputation from internal cache |
| iprepd.err.timeout | count | Request to iprepd timed out |
| iprepd.err.500 | count | Got a 500 response from iprepd |
| iprepd.err.401 | count | Got a 401 response from iprepd, usually means the API key in use is invalid or being sent incorrectly by nginx. |
| iprepd.err.dns_timeout | count | DNS resolution of the iprepd URL's domain name timed out. Make sure to check nginx's [resolver_timeout](https://nginx.org/en/docs/http/ngx_http_core_module.html#resolver_timeout) setting |
| iprepd.err.* | count | Got an error while sending a request to iprepd. This could be other 4xx or 5xx status codes for example. |


#### Setting up custom metrics

You can use `client.statsd` (where `client = require("resty.iprepd").new({...})`) to submit your
own custom metrics. Do note that there is no prefix, so it will act as any other statsd client.

##### Available statsd functions

```lua
client = require("resty.iprepd").new({...})

client.statsd.count(name, value)
client.statsd.incr(name) # Increments a count by 1
client.statsd.time(name, value)
client.statsd.set(name, value)
```

##### Example within nginx config
```
init_by_lua_block {
  client = require("resty.iprepd").new({
    url = os.getenv("IPREPD_URL"),
    api_key = os.getenv("IPREPD_API_KEY"),
    statsd_host = os.getenv("STATSD_HOST"),
  })
}

init_worker_by_lua_block {
  # async flushing of metrics
  client:config_flush_timer()
}

server {
  ...

  location / {
    ...

    access_by_lua_block {
      client:check(ngx.var.remote_addr)
    }

    log_by_lua_block {
      # This conditional is not required, but can be helpful to not cause problems
      # if you want to temporarily disable statsd. This will evaluate to false if
      # `statsd_host` is not set.
      if client.statsd then
        # Here is our custom metric
        client.statsd.set("iprepd.ips_seen", ngx.var.remote_addr)
      end
    }
  }
}
```

### Common Gotchas

* Make sure iprepd-nginx is seeing the real client IP. You will usually need to use something like [ngx_http_realip_module](https://nginx.org/en/docs/http/ngx_http_realip_module.html), and confirm that it is configured correctly.

## Running iprepd-nginx dev environment locally
A dev environment is provided that can be used to test iprepd-nginx with iprepd and a dummy backend service (in this case nginx with static content).

```
$ make build
$ make run_dev_env
```

Sanity tests can then be run by:
```
$ make build_test
$ make smoke_test
```

## Running just iprepd-nginx locally
Create a `.env` file in this repo with the needed environment variables (documentation below).

Then run:
```
$ make build
$ make run_dev
```

Then you will be able to hit this proxy with: `curl http://localhost:80`

### Environment Variables for Dev

#### Note:

Quotations in env vars matter with nginx. Don't use them if you are using `--env-file` in Docker.

```
# required
backend=http://<>               # URL to proxy to
IPREPD_URL=http://<>            # iprepd url
IPREPD_API_KEY="api-key"        # iprepd api key
IPREPD_REPUTATION_THRESHOLD=50  # iprepd reputation threshold, block all IP's with a reputation below the threshold

#
# optional
#
IPREPD_TIMEOUT=10
IPREPD_CACHE_TTL=30
IPREPD_CACHE_ERRORS=0
IPREPD_CACHE_BUFFER_COUNT=5000
IPREPD_WHITELISTED_LIST=10.0.0.0/8,127.0.0.1/32
STATSD_HOST=127.0.0.1
STATSD_PORT=8125
STATSD_MAX_BUFFER_COUNT=200
STATSD_FLUSH_TIMER=2
BLOCKING_MODE=0
AUDIT_BLOCKED_REQUESTS=0
AUDIT_INCLUDE_HEADERS=0
AUDIT_URI_LIST=/uri1,/uri2
```


CHANGELOG.md
## Change log

### 0.1.14

* `IPREPD_WHITELISTED_LIST` has been deprecated and `IPREPD_ALLOWLIST` has been added. If both are specified `IPREPD_ALLOWLIST` takes precedence. `IPREPD_WHITELISTED_LIST` will be removed in a future release.
* Build and test process has changed resulting in leaner docker image for production deployments
* The allowlist functionality now supports ipv6. This was done by changing a dependency from iputils to ipmatcher.
* Uses newer iprepd api version
### 0.1.13

* Expect comma delimited strings for `IPREPD_WHITELISTED_LIST` and `AUDIT_URI_LIST` and parse
  them in the module, instead of requiring tables be constructed in the nginx `init_by_lua_block`


test/test.sh
#!/bin/bash

source /opt/iprepd-nginx/etc/test-env

/usr/bin/redis-server --daemonize yes
while ! nc -z localhost 6379; do sleep 0.1; done

nohup /root/go/bin/iprepd -c /opt/iprepd-nginx/etc/iprepd.yaml >/dev/null 2>&1 &
while ! nc -z localhost 8080; do sleep 0.1; done

(cd /opt/iprepd-nginx/test && pytest -s)


test/test_module.py
#!/usr/bin/env python3

import socketserver
import http.server
import subprocess
import json
import requests
import time
import pytest
import threading
import os

class openresty_runner(threading.Thread):
    cmd = ['/usr/bin/openresty', '-g', 'daemon off;']

    def run(self):
        self.proc = subprocess.Popen(self.cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE)

    def stop(self):
        self.proc.terminate()
        out, err = self.proc.communicate()
        return out, err

    def begin(self):
        self.start()
        # Wait until the daemon is up and running
        while True:
            try:
                r = requests.get('http://127.0.0.1/health')
                break
            except Exception as e:
                time.sleep(0.1)

    def __init__(self, conf=None):
        if conf:
            self.cmd = self.cmd + ['-c', conf]
        threading.Thread.__init__(self)

class iprepd_mock(threading.Thread):
    mode_delay = 1
    mode_error = 2
    mode_ok = 3

    def run(self):
        self.httpd.serve_forever()

    def shutdown(self):
        self.httpd.shutdown()

    def __init__(self):
        threading.Thread.__init__(self)
        self.mode = self.mode_error
        socketserver.ThreadingTCPServer.allow_reuse_address = True
        self.httpd = socketserver.ThreadingTCPServer(('', 8081), http_handler)

# XXX This global and specifically the use of the mode field to toggle behavior isn't
# great and should be replaced with something more precise.
iprepd_mock_thread = None

class http_handler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        if iprepd_mock_thread.mode == iprepd_mock.mode_error:
            self.send_response(500)
            self.end_headers()
        elif iprepd_mock_thread.mode == iprepd_mock.mode_delay:
            time.sleep(5)
            self.send_response(404)
            self.end_headers()
        elif iprepd_mock_thread.mode == iprepd_mock.mode_ok:
            buf = json.dumps({
                'object': '127.0.0.1',
                'type': 'ip',
                'reputation': 25
            })
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(buf.encode(encoding='utf_8'))

@pytest.fixture(scope='session', autouse=True)
def start_http(request):
    global iprepd_mock_thread
    iprepd_mock_thread = iprepd_mock()
    request.addfinalizer(stop_http)
    iprepd_mock_thread.start()

def stop_http():
    iprepd_mock_thread.shutdown()

def test_defaults():
    # Reset IP reputation as part of fixture before test
    update_reputation(100, '127.0.0.1')

@pytest.fixture
def openresty():
    test_defaults()
    return openresty_runner()

@pytest.fixture
def openresty_rl():
    test_defaults()
    return openresty_runner(conf='/opt/iprepd-nginx/etc/testconf/rl/nginx.conf')

def update_reputation(rep, ip):
    body = { 'object': ip, 'type': 'ip', 'reputation': rep }
    headers = { 'content-type': 'application/json',
        'authorization': 'APIKey ' + os.environ['IPREPD_API_KEY'] }
    requests.put(os.environ['IPREPD_URL'] + '/type/ip/127.0.0.1',
        json=body, headers=headers)

def simple_request():
    return requests.get('http://127.0.0.1/iprepd_ping')

def post_request():
    return requests.post('http://127.0.0.1/iprepd_ping', data='body')

def test_simple_request(openresty):
    openresty.begin()
    ret = simple_request()
    _, err = openresty.stop()
    assert ret.status_code == 200
    assert ret.text == 'pong\n'
    assert len(err) == 0

def test_bad_apikey(openresty):
    oldkey = os.environ['IPREPD_API_KEY']
    os.environ['IPREPD_API_KEY'] = 'invalid'
    openresty.begin()
    ret = simple_request()
    _, err = openresty.stop()
    assert ret.status_code == 200
    assert 'iprepd responded with a 401 http status code' in str(err)
    os.environ['IPREPD_API_KEY'] = oldkey

def test_reputation_fail(openresty):
    update_reputation(0, '127.0.0.1')
    os.environ['BLOCKING_MODE'] = '1'
    openresty.begin()
    ret = simple_request()
    _, err = openresty.stop()
    assert ret.status_code == 429
    assert '127.0.0.1 rejected with a reputation of 0' in str(err)
    del os.environ['BLOCKING_MODE']

def test_serial_range(openresty):
    update_reputation(0, '127.0.0.1')
    os.environ['BLOCKING_MODE'] = '1'
    # Set the smallest possible cache size for the test
    os.environ['IPREPD_CACHE_BUFFER_COUNT'] = '1'
    openresty.begin()
    for _ in range(250):
        ret = simple_request()
        assert ret.status_code == 429
    _, err = openresty.stop()
    del os.environ['BLOCKING_MODE']
    del os.environ['IPREPD_CACHE_BUFFER_COUNT']

def test_serial_range_rl(openresty_rl):
    update_reputation(0, '127.0.0.1')
    os.environ['BLOCKING_MODE'] = '1'
    # Set the smallest possible cache size for the test
    os.environ['IPREPD_CACHE_BUFFER_COUNT'] = '1'
    openresty_rl.begin()
    for _ in range(250):
        ret = simple_request()
        assert ret.status_code == 429
    _, err = openresty_rl.stop()
    del os.environ['BLOCKING_MODE']
    del os.environ['IPREPD_CACHE_BUFFER_COUNT']

def test_reputation_above(openresty):
    update_reputation(51, '127.0.0.1')
    os.environ['BLOCKING_MODE'] = '1'
    openresty.begin()
    ret = simple_request()
    _, err = openresty.stop()
    assert ret.status_code == 200
    assert len(err) == 0
    del os.environ['BLOCKING_MODE']

def test_cache_ttl(openresty):
    os.environ['BLOCKING_MODE'] = '1'
    os.environ['IPREPD_CACHE_TTL'] = '3'
    openresty.begin()
    ret = simple_request()
    assert ret.status_code == 200
    update_reputation(0, '127.0.0.1')
    ret = simple_request()
    # Should still be in cache as OK
    assert ret.status_code == 200
    # Sleep to let the cache entry expire
    time.sleep(4)
    ret = simple_request()
    assert ret.status_code == 429
    _, err = openresty.stop()
    assert len(str(err).split('\n')) == 1
    del os.environ['BLOCKING_MODE']
    del os.environ['IPREPD_CACHE_TTL']

def test_request_timeout(openresty):
    # Make use of simulated inoperable service
    oldurl = os.environ['IPREPD_URL']
    os.environ['IPREPD_URL'] = 'http://127.0.0.1:8081'
    iprepd_mock_thread.mode = iprepd_mock.mode_delay
    openresty.begin()
    ret = simple_request()
    _, err = openresty.stop()
    assert ret.status_code == 200
    assert 'lua tcp socket read timed out' in str(err)
    os.environ['IPREPD_URL'] = oldurl

def test_cache_errors_disabled(openresty):
    os.environ['BLOCKING_MODE'] = '1'
    # Make use of simulated inoperable service
    oldurl = os.environ['IPREPD_URL']
    os.environ['IPREPD_URL'] = 'http://127.0.0.1:8081'
    iprepd_mock_thread.mode = iprepd_mock.mode_error
    openresty.begin()
    ret = simple_request()
    # Should have resulted in an error, so we'd get a 200 back
    assert ret.status_code == 200
    iprepd_mock_thread.mode = iprepd_mock.mode_ok
    ret = simple_request()
    # Error would not have been cached so should result in new check
    assert ret.status_code == 429
    _, err = openresty.stop()
    os.environ['IPREPD_URL'] = oldurl
    del os.environ['BLOCKING_MODE']

def test_cache_errors_enabled(openresty):
    os.environ['BLOCKING_MODE'] = '1'
    os.environ['IPREPD_CACHE_ERRORS'] = '1'
    # Make use of simulated inoperable service
    oldurl = os.environ['IPREPD_URL']
    os.environ['IPREPD_URL'] = 'http://127.0.0.1:8081'
    iprepd_mock_thread.mode = iprepd_mock.mode_error
    openresty.begin()
    ret = simple_request()
    # Should have resulted in an error, so we'd get a 200 back
    assert ret.status_code == 200
    iprepd_mock_thread.mode = iprepd_mock.mode_ok
    ret = simple_request()
    # Should have been cached so will still get a 200
    assert ret.status_code == 200
    _, err = openresty.stop()
    os.environ['IPREPD_URL'] = oldurl
    del os.environ['BLOCKING_MODE']
    del os.environ['IPREPD_CACHE_ERRORS']

def test_audit_requests_get(openresty):
    os.environ['BLOCKING_MODE'] = '1'
    os.environ['AUDIT_BLOCKED_REQUESTS'] = '1'
    os.environ['AUDIT_URI_LIST'] = '/iprepd_ping'
    update_reputation(0, '127.0.0.1')
    openresty.begin()
    ret = simple_request()
    assert ret.status_code == 429
    _, err = openresty.stop()
    assert 'audit_log(): FoxSec Audit || ip: 127.0.0.1, headers: , body: ' + \
        'unable to read body, client: 127.0.0.1, server: , request: "GET ' + \
        '/iprepd_ping HTTP/1.1", host: "127.0.0.1"' in str(err)
    del os.environ['BLOCKING_MODE']
    del os.environ['AUDIT_BLOCKED_REQUESTS']
    del os.environ['AUDIT_URI_LIST']

def test_audit_requests_get_pattern(openresty):
    os.environ['BLOCKING_MODE'] = '1'
    os.environ['AUDIT_BLOCKED_REQUESTS'] = '1'
    os.environ['AUDIT_URI_LIST'] = '/something,/iprepd_%l%l%l%l'
    update_reputation(0, '127.0.0.1')
    openresty.begin()
    ret = simple_request()
    assert ret.status_code == 429
    _, err = openresty.stop()
    assert 'audit_log(): FoxSec Audit || ip: 127.0.0.1, headers: , body: ' + \
        'unable to read body, client: 127.0.0.1, server: , request: "GET ' + \
        '/iprepd_ping HTTP/1.1", host: "127.0.0.1"' in str(err)
    del os.environ['BLOCKING_MODE']
    del os.environ['AUDIT_BLOCKED_REQUESTS']
    del os.environ['AUDIT_URI_LIST']

def test_audit_requests_get_pattern(openresty):
    os.environ['BLOCKING_MODE'] = '1'
    os.environ['AUDIT_BLOCKED_REQUESTS'] = '1'
    os.environ['AUDIT_URI_LIST'] = '/something,/iprepd_%l%l%l%l'
    update_reputation(0, '127.0.0.1')
    openresty.begin()
    ret = post_request()
    assert ret.status_code == 429
    _, err = openresty.stop()
    assert 'audit_log(): FoxSec Audit || ip: 127.0.0.1, headers: , body: Ym9keQ==' + \
        ', client: 127.0.0.1, server: , request: "POST /iprepd_ping HTTP/1.1", ho' + \
        'st: "127.0.0.1"' in str(err)
    del os.environ['BLOCKING_MODE']
    del os.environ['AUDIT_BLOCKED_REQUESTS']
    del os.environ['AUDIT_URI_LIST']

def test_whitelist_match_multi(openresty):
    update_reputation(0, '127.0.0.1')
    os.environ['BLOCKING_MODE'] = '1'
    os.environ['IPREPD_WHITELISTED_LIST'] = '10.0.0.0/8,127.0.0.0/24'
    openresty.begin()
    ret = simple_request()
    _, err = openresty.stop()
    assert ret.status_code == 200
    del os.environ['BLOCKING_MODE']
    del os.environ['IPREPD_WHITELISTED_LIST']

def test_whitelist_match_single(openresty):
    update_reputation(0, '127.0.0.1')
    os.environ['BLOCKING_MODE'] = '1'
    os.environ['IPREPD_WHITELISTED_LIST'] = '127.0.0.1/32'
    openresty.begin()
    ret = simple_request()
    _, err = openresty.stop()
    assert ret.status_code == 200
    del os.environ['BLOCKING_MODE']
    del os.environ['IPREPD_WHITELISTED_LIST']

def test_whitelist_nomatch_single(openresty):
    update_reputation(0, '127.0.0.1')
    os.environ['BLOCKING_MODE'] = '1'
    os.environ['IPREPD_WHITELISTED_LIST'] = '127.0.0.2/32'
    openresty.begin()
    ret = simple_request()
    _, err = openresty.stop()
    assert ret.status_code == 429
    del os.environ['BLOCKING_MODE']
    del os.environ['IPREPD_WHITELISTED_LIST']


lib/resty/statsd.lua
--
-- Forked from https://github.com/lonelyplanet/openresty-statsd
--
local _M = {}

_M.buffer = {count=0}

function _M.flush(host, port)
  if pcall(function()
    local udp = ngx.socket.udp()
    udp:setpeername(host, port)
    _M.buffer['count'] = nil
    udp:send(_M.buffer)
    udp:close()
  end) then
    -- pass
  else
    ngx.log(ngx.ERR, 'Error sending stats to statsd at ' .. host .. ':' .. port)
  end

  -- reset buffer
  _M.buffer = {count=0}
end

function _M.buffer_count()
  return _M.buffer['count']
end

function _M.register(bucket, suffix)
  _M.buffer['count'] = _M.buffer['count'] + 1
  table.insert(_M.buffer, bucket .. ':' .. suffix .. '\n')
end

function _M.time(bucket, time)
  _M.register(bucket, time .. '|ms')
end

function _M.set(bucket, value)
  _M.register(bucket, value .. '|s')
end

function _M.count(bucket, n)
  _M.register(bucket, n .. '|c')
end

function _M.incr(bucket)
  _M.count(bucket, 1)
end

return _M


lib/resty/iprepd.lua
local cjson = require('cjson')
local http = require('resty.http')
local lrucache = require('resty.lrucache')
local statsd = require('resty.statsd')
local ipmatcher = require('resty.ipmatcher')

function fatal_error(error_msg)
  ngx.log(ngx.ERR, error_msg)
  os.exit(1)
end

function warning(message)
  ngx.log(ngx.ERR, message)
end

local _M = {}
local mt = { __index = _M }

function _M.new(options)
  local iprepd_url = options.url or 'http://localhost:8080'

  if iprepd_url:sub(-1) == '/' then
    iprepd_url = iprepd_url:sub(1, -2)
  end

  local cache_buffer_count = options.cache_buffer_count or 5000
  local iprepd_threshold = options.threshold or fatal_error('Need to pass in a threshold')
  local iprepd_api_key = options.api_key or fatal_error('Need to pass in an api_key')

  local cache, err = lrucache.new(cache_buffer_count)
  if not cache then
    fatal_error(string.format('failed to create the cache: %s', (err or 'unknown')))
  end

  local statsd_client = nil
  if options.statsd_host then
    statsd_client = statsd
  end

  local audit_uri_list = {}
  if options.audit_uri_list and type(options.audit_uri_list) == 'string' then
    for s in string.gmatch(options.audit_uri_list, "([^,]+)") do
      table.insert(audit_uri_list, s)
    end
  end

  local allowlist = nil
  local allowlist_table = {}
  if options.whitelist and type(options.whitelist) == 'string' then
    warning("whitelist option is deprecated. please switch to allowlist.")
    for s in string.gmatch(options.whitelist, "([^,]+)") do
      table.insert(allowlist_table, s)
    end
    allowlist = ipmatcher.new(allowlist_table)
  end

  if options.allowlist and type(options.allowlist) == 'string' then
    if type(next(allowlist_table)) ~= "nil" then
      warning("both whitelist and allowlist specified - using allowlist")
      allowlist_table = {}
    end
    for s in string.gmatch(options.allowlist, "([^,]+)") do
      table.insert(allowlist_table, s)
    end
    allowlist = ipmatcher.new(allowlist_table)
  end


  if options.audit_blocked_requests == 1 then
    if next(audit_uri_list) == nil then
      fatal_error("audit blocked requests is enabled but no uris are specified")
    end
  end

  local self = {
    url = iprepd_url,
    timeout = options.timeout or 10,
    threshold = iprepd_threshold,
    iprepd_hdrs = {
      ['Authorization'] = string.format('APIKey %s', iprepd_api_key),
      ['Connection'] = 'keep-alive',
    },
    cache = cache,
    cache_ttl = options.cache_ttl or 60,
    cache_errors = options.cache_errors or 0,
    cache_errors_ttl = options.cache_errors_ttl or 10,
    statsd = statsd_client,
    statsd_host = options.statsd_host,
    statsd_port = options.statsd_port or 8125,
    statsd_max_buffer_count =  options.statsd_max_buffer_count or 100,
    statsd_flush_timer = options.statsd_flush_timer or 5,
    blocking_mode = options.blocking_mode or 0,
    verbose = options.verbose or 0,
    allowlist = allowlist,
    audit_blocked_requests = options.audit_blocked_requests or 0,
    audit_include_headers = options.audit_include_headers or 0,
    audit_uri_list = audit_uri_list,
  }

  return setmetatable(self, mt)
end

function _M.check(self, ip)
  self:debug_log(string.format("Checking %s", ip))
  ngx.req.set_header('X-Foxsec-IP-Reputation-Below-Threshold', 'false')
  ngx.req.set_header('X-Foxsec-Block', 'false')
  
  if self.allowlist then
    if self.allowlist:match(ip) then
      self:debug_log(string.format("%s in allowlist", ip))
      return
    end
  end

  local reputation = self:get_reputation(ip)
  if reputation then
    self:debug_log(string.format("Got reputation of %d for %s", reputation, ip))
    ngx.req.set_header('X-Foxsec-IP-Reputation', tostring(reputation))
    if reputation <= self.threshold then
      ngx.req.set_header('X-Foxsec-IP-Reputation-Below-Threshold', 'true')
      ngx.req.set_header('X-Foxsec-Block', 'true')

      self:audit_log(ip)
      if self.statsd then
        self.statsd.incr("iprepd.status.below_threshold")
      end

      if self.blocking_mode == 0 then
        ngx.log(ngx.ERR, string.format("%s is below threshold with a reputation of %d", ip, reputation))
      else
        ngx.log(ngx.ERR, string.format("%s rejected with a reputation of %d", ip, reputation))
        if self.statsd then
          self.statsd.incr("iprepd.status.rejected")
        end
        -- respond with a 429
        ngx.exit(ngx.HTTP_TOO_MANY_REQUESTS)
      end
    end
  end

  self:debug_log(string.format("%s accepted", ip))
  if self.statsd then
    self.statsd.incr("iprepd.status.accepted")
  end
end

function _M.get_reputation(self, ip)
  local reputation = self.cache:get(ip)

  if not reputation then
    local httpc = http.new()
    httpc:set_timeout(self.timeout)
    local resp, err = httpc:request_uri(string.format("%s/type/ip/%s", self.url, ip), {
      method  = "GET",
      headers = self.iprepd_hdrs,
    })
    self.statsd.incr("iprepd.get_reputation")
    if err then
      if self.statsd then
        if string.find(err, " ") then
          if string.find(err, "could not be resolved") and string.find(err, "Operation timed out") then
            self.statsd.incr("iprepd.err.dns_timeout")
          else
            ngx.log(ngx.ERR, string.format("Could not send metric with error: %s", err))
          end
        else
          self.statsd.incr("iprepd.err." .. err)
        end
      end
      ngx.log(ngx.ERR, string.format("Error with request to iprepd: %s", err))
      return nil
    end

    -- If the IP was found
    if resp.status == 200 then
      reputation = cjson.decode(resp.body)['reputation']
      if not reputation then
        ngx.log(ngx.ERR, 'Unable to parse `reputation` value from response body')
      end
    elseif resp.status == 404 then
      reputation = 100
      self:debug_log(string.format("no reputation for %s", ip))
    else
      ngx.log(ngx.ERR, string.format("iprepd responded with a %d http status code", resp.status))
      if self.statsd then
        self.statsd.incr("iprepd.err." .. resp.status)
      end
      if self.cache_errors == 1 then
        reputation = 100
        self:debug_log(string.format("cache_errors is enabled, setting reputation of %s to 100 within the cache", ip))
        self.cache:set(ip, reputation, self.cache_errors_ttl)
        return reputation
      end
    end

    if reputation and reputation >= 0 and reputation <= 100 then
      self.cache:set(ip, reputation, self.cache_ttl)
    end
  else
    self.statsd.incr("iprepd.cache_hit")
  end

  return reputation
end

function _M.flush_stats(self)
  if self.statsd then
    if self.statsd.buffer_count() >= self.statsd_max_buffer_count then
      self.statsd.flush(self.statsd_host, self.statsd_port)
    end
  end
end

function _M.async_flush_stats(premature, self)
  if self.statsd then self.statsd.flush(self.statsd_host, self.statsd_port) end
end

function _M.config_flush_timer(self)
  ngx.timer.every(self.statsd_flush_timer, self.async_flush_stats, self)
end

function _M.debug_log(self, msg)
  if self.verbose == 1 then
    ngx.log(ngx.ERR, string.format("[verbose] %s", msg))
  end
end

function _M.audit_log(self, ip)
  if self.audit_blocked_requests == 1 then
    local should_audit_request = 0
    local headers = ngx.req.get_headers()
    for i, uri in ipairs(self.audit_uri_list) do
      if string.match(ngx.var.uri, uri) then
        should_audit_request = 1
        break
      end
    end
    if should_audit_request == 1 then
      ngx.req.read_body()
      local request_headers_all = ""
      if self.audit_include_headers == 1 then
        local headers = ngx.req.get_headers()
        headers['authorization'] = nil
        headers['proxy-authorization'] = nil
        for k,v in pairs(headers) do
          request_headers_all = request_headers_all .. string.format('"%s": "%s", ', k, v)
        end
      end
      request_headers_all = ngx.encode_base64(request_headers_all)
      local data = ngx.req.get_body_data()
      if data then
        data = ngx.encode_base64(data)
      end
      if not data then
        data = ngx.req.get_body_file()
      end
      if not data then
        data = "unable to read body"
      end
      ngx.log(ngx.ALERT, string.format('FoxSec Audit || ip: %s, headers: %s, body: %s', ip, request_headers_all, data))
    end
  end
end

return _M
